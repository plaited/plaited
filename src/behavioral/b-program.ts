import type { BPEvent, BPEventTemplate, BPListener, Idioms, RulesFunction } from './b-thread.js'
import { isTypeOf } from '../utils/is-type-of.js'

type RunningBid = {
  trigger?: true | 'object' | 'person'
  priority: number
  generator: IterableIterator<Idioms>
}
type PendingBid = Idioms & RunningBid

type CandidateBid = {
  thread: string
  priority: number
  type: string
  detail?: unknown
  trigger?: true | 'object' | 'person'
  template?: BPEventTemplate
}

export type Disconnect = () => void

export type SnapshotMessage = {
  thread: string
  selected: boolean
  type: string
  detail?: unknown
  priority: number
  blockedBy?: string
}[]

type SnapshotFormatter = (args: {
  pending: Map<string, PendingBid>
  selectedEvent: CandidateBid
  candidates: CandidateBid[]
}) => SnapshotMessage

export type SnapshotListener = (msg: SnapshotMessage) => void | Promise<void>

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type DefaultHandlers = Record<string, (detail: any) => void | Promise<void>>

export type Handlers<T = DefaultHandlers> = DefaultHandlers & T
export type UseFeedback = <T = DefaultHandlers>(handlers: Handlers<T>) => Disconnect
export type UseSnapshot = (listener: SnapshotListener) => Disconnect
export type BThreads = {
  has: (thread: string) => { running: boolean; pending: boolean }
  set: (threads: Record<string, RulesFunction>) => void
}

export type Trigger = <T>(args: BPEvent<T>) => void

export type BProgram = () => Readonly<{
  bThreads: BThreads
  trigger: Trigger
  useFeedback: UseFeedback
  useSnapshot: UseSnapshot
}>

const triggerWaitFor = () => true

const createPublisher = <T>() => {
  const listeners = new Set<(value: T) => void | Promise<void>>()
  function publisher(value: T) {
    for (const cb of listeners) {
      void cb(value)
    }
  }
  publisher.subscribe = (listener: (msg: T) => void | Promise<void>) => {
    listeners.add(listener)
    return () => {
      listeners.delete(listener)
    }
  }
  return publisher
}

const ensureArray = <T>(obj: T | T[] = []) => (Array.isArray(obj) ? obj : [obj])

const isListeningFor = ({ type, detail }: CandidateBid) => {
  return (listener: BPListener): boolean =>
    typeof listener !== 'string' ?
      listener({
        detail,
        type,
      })
    : listener === type
}

const isPendingRequest = (selectedEvent: CandidateBid, event: BPEvent | BPEventTemplate) =>
  isTypeOf<BPEventTemplate>(event, 'function') ? event === selectedEvent?.template : event.type == selectedEvent.type

const snapshotFormatter: SnapshotFormatter = ({ candidates, selectedEvent, pending }) => {
  const blockingThreads = [...pending].flatMap(([thread, { block }]) =>
    block && Array.isArray(block) ? block.map((listener) => ({ block: listener, thread }))
    : block ? [{ block, thread }]
    : [],
  )
  const ruleSets: {
    thread: string
    selected: boolean
    type: string
    detail?: unknown
    priority: number
    blockedBy?: string
    trigger?: true | 'object' | 'person'
  }[] = []
  for (const bid of candidates) {
    const blockedCB = isListeningFor(bid)
    ruleSets.push({
      thread: bid.thread,
      selected: isPendingRequest(selectedEvent, bid),
      type: bid.type,
      priority: bid.priority,
      detail: bid.detail,
      blockedBy: blockingThreads.find(({ block }) => blockedCB(block))?.thread,
      trigger: bid.trigger,
    })
  }
  return ruleSets.sort((a, b) => a.priority - b.priority)
}
/**
 * Creates a behavioral program that manages the execution of behavioral threads.
 */
export const bProgram: BProgram = () => {
  const pending = new Map<string, PendingBid>()
  const running = new Map<string, RunningBid>()

  const actionPublisher = createPublisher<BPEvent>()
  let snapshotPublisher:
    | {
        (value: SnapshotMessage): void
        subscribe(listener: (msg: SnapshotMessage) => void | Promise<void>): () => void
      }
    | undefined
  function run() {
    running.size && step()
  }

  function step() {
    for (const [thread, bid] of running) {
      const { generator, priority, trigger } = bid
      const { value, done } = generator.next()
      !done &&
        pending.set(thread, {
          priority,
          ...(trigger && { trigger }),
          generator,
          ...value,
        })
      running.delete(thread)
    }
    selectNextEvent()
  }
  // Select next event
  function selectNextEvent() {
    const blocked: BPListener[] = []
    const candidates: CandidateBid[] = []
    for (const [thread, { request, priority, block, trigger }] of pending) {
      block && blocked.push(...ensureArray(block))
      request &&
        candidates.push({
          priority,
          trigger,
          thread,
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          ...(isTypeOf<BPEventTemplate<any>>(request, 'function') ? { template: request, ...request() } : request),
        })
    }
    const filteredBids: CandidateBid[] = []
    const length = candidates.length
    for (let i = 0; i < length; i++) {
      const candidate = candidates[i]
      // Are we blocking the the candidate event
      if (!blocked.some(isListeningFor(candidate))) {
        filteredBids.push(candidate)
      }
    }
    /** @description Priority Queue BPEvent Selection Strategy */
    const selectedEvent = filteredBids.sort(
      ({ priority: priorityA }, { priority: priorityB }) => priorityA - priorityB,
    )[0]
    if (selectedEvent) {
      snapshotPublisher && snapshotPublisher(snapshotFormatter({ candidates, selectedEvent, pending }))
      nextStep(selectedEvent)
    }
  }
  // Queue up bids for next step of super step
  function nextStep(selectedEvent: CandidateBid) {
    for (const [thread, bid] of pending) {
      const { waitFor, request, generator, interrupt } = bid
      const isInterrupted = ensureArray(interrupt).some(isListeningFor(selectedEvent))
      const isWaitedFor = ensureArray(waitFor).some(isListeningFor(selectedEvent))
      const hasPendingRequest = request && isPendingRequest(selectedEvent, request)
      isInterrupted && generator.return?.()
      if (hasPendingRequest || isInterrupted || isWaitedFor) {
        running.set(thread, bid)
        pending.delete(thread)
      }
    }
    // To avoid infinite loop with calling trigger from feedback always publish select event
    // after checking if request(s) is waitList and before our next run
    actionPublisher({ type: selectedEvent.type, detail: selectedEvent.detail })
    run()
  }

  const trigger: Trigger = (request) => {
    const thread = function* () {
      yield {
        request,
        waitFor: [triggerWaitFor],
      }
    }
    running.set(request.type, {
      priority: 0,
      trigger: true,
      generator: thread(),
    })
    run()
  }

  const useFeedback: UseFeedback = (handlers) => {
    const disconnect = actionPublisher.subscribe((data: BPEvent) => {
      const { type, detail = {} } = data
      if (Object.hasOwn(handlers, type)) {
        void handlers[type](detail)
      }
    })
    return disconnect
  }

  const bThreads: BThreads = {
    set: (threads) => {
      for (const thread in threads) {
        running.set(thread, {
          priority: running.size + 1,
          generator: threads[thread](),
        })
      }
    },
    has: (thread) => ({ running: running.has(thread), pending: pending.has(thread) }),
  }

  const useSnapshot: UseSnapshot = (listener) => {
    if (snapshotPublisher === undefined) snapshotPublisher = createPublisher<SnapshotMessage>()
    return snapshotPublisher.subscribe(listener)
  }

  return Object.freeze({
    /** add threads functions to behavioral program and verify if there still active */
    bThreads,
    /** trigger a run and event on behavioral program */
    trigger,
    /** connect action function to behavioral program */
    useFeedback,
    /** connect a listener that receives state snapshots for each step of a running behavioral program */
    useSnapshot,
  })
}
