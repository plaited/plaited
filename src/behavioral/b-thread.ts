import { isTypeOf } from '../utils/is-type-of.js'

/**
 * Defines the repetition behavior for a `bThread`.
 * - `true`: The thread repeats indefinitely.
 * - `function`: A predicate function evaluated before each repetition. The thread repeats if the function returns `true`.
 */
type Repeat = true | (() => boolean)

/**
 * Represents a fundamental unit of communication in behavioral programming.
 * An event consists of a mandatory `type` (string identifier) and an optional `detail` payload.
 *
 * @template T The expected type of the `detail` payload. Defaults to `unknown`.
 * @property type - The string identifier for the event.
 * @property detail - Optional data associated with the event.
 * @example
 * const userLoginEvent: BPEvent<{ userId: string }> = { type: 'USER_LOGIN', detail: { userId: 'usr123' } };
 * const simpleEvent: BPEvent = { type: 'SYSTEM_START' };
 */
export type BPEvent<T = unknown> = { type: string; detail?: T }

/**
 * A factory function that generates a `BPEvent`.
 * Useful within `bSync` definitions when the event details might be computed dynamically
 * at the time the synchronization point is reached.
 *
 * @template T The expected type of the `detail` payload for the generated event. Defaults to `unknown`.
 * @returns A `BPEvent<T>` object.
 * @example
 * const createTimestampedEvent = (type: string): BPEventTemplate<{ timestamp: number }> =>
 *   () => ({ type, detail: { timestamp: Date.now() } });
 *
 * bSync({ request: createTimestampedEvent('LOG_TIME') });
 */
export type BPEventTemplate<T = unknown> = () => BPEvent<T>

/**
 * Defines how a b-thread listens for or specifies events in `waitFor`, `block`, or `interrupt` idioms.
 * It can be:
 * 1. A simple `string`: Matches events exactly by their `type`.
 * 2. A predicate function: Takes an event object (`{ type: string, detail: T }`) and returns `true` if the event matches the desired criteria.
 *
 * @template T The expected type of the `detail` payload for events being listened for. Defaults to `unknown`.
 * @example
 * // Listen for event type 'USER_ACTION'
 * const listener1: BPListener = 'USER_ACTION';
 *
 * // Listen for any event whose type starts with 'UI_'
 * const listener2: BPListener = ({ type }) => type.startsWith('UI_');
 *
 * // Listen for 'DATA_LOADED' events with a specific status in the detail
 * const listener3: BPListener<{ status: string }> = ({ type, detail }) =>
 *   type === 'DATA_LOADED' && detail?.status === 'success';
 *
 * bSync({ waitFor: [listener1, listener2], block: listener3 });
 */
export type BPListener<T = unknown> = string | ((args: { type: string; detail: T }) => boolean)

/**
 * Represents a synchronization statement within a b-thread's generator function (`RulesFunction`).
 * It defines the thread's intentions at a specific point in its execution, specifying which events it wants to:
 * - `request`: Propose an event to be selected and triggered by the `bProgram`. Only one request can be active per sync point.
 * - `waitFor`: Wait for specific events to occur. The thread pauses until a matching event is selected.
 * - `block`: Prevent specific events from being selected by the `bProgram` while this sync point is active.
 * - `interrupt`: Specify events that, if selected, will cause the thread to terminate its current execution flow and potentially restart if configured to repeat.
 *
 * Multiple listeners can be provided for `waitFor`, `block`, and `interrupt` using arrays.
 *
 * @template T The expected type of the `detail` payload for events involved in this synchronization point. Defaults to `any` for flexibility, but specific types are recommended.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Idioms<T = any> = {
  /** Event(s) the thread is waiting for. Execution pauses until a matching event is selected. */
  waitFor?: BPListener<T> | BPListener<T>[]
  /** Event(s) that will interrupt the thread's execution if selected. */
  interrupt?: BPListener<T> | BPListener<T>[]
  /** An event the thread wishes to request. Can be a static event object or a template function. */
  request?: BPEvent<T> | BPEventTemplate<T>
  /** Event(s) the thread wants to prevent from being selected. */
  block?: BPListener<T> | BPListener<T>[]
}

/**
 * A generator function defining the behavior of a b-thread.
 * It yields `Idioms` objects, representing synchronization points where the thread
 * interacts with the `bProgram` and other threads. The generator pauses at each `yield`
 * and resumes when the `bProgram` selects an event relevant to its `waitFor` or `interrupt` declarations,
 * or when its `request` is selected.
 *
 * @returns A Generator that yields `Idioms` and ultimately returns `void`.
 * @example
 * function* myThreadRules(): RulesFunction {
 *   yield { request: { type: 'INIT' } }; // Request INIT event
 *   yield { waitFor: 'INIT_COMPLETE' }; // Wait for INIT_COMPLETE
 *   console.log('Initialization complete!');
 * }
 */
export type RulesFunction = () => Generator<Idioms, void, undefined>

/**
 * A factory function that creates a single synchronization step (a `RulesFunction`) for a b-thread.
 * It takes an `Idioms` object describing the desired synchronization behavior (request, waitFor, block, interrupt).
 *
 * @template T The type of the event detail payload relevant to this synchronization point.
 * @param arg The `Idioms<T>` object defining the synchronization behavior.
 * @returns A `RulesFunction` (generator function) that yields the provided `Idioms` object once.
 */
export type BSync = <T>(arg: Idioms<T>) => () => Generator<Idioms, void, unknown>

/**
 * A factory function that constructs a complete b-thread (`RulesFunction`) by composing multiple synchronization steps.
 *
 * @param rules An array of `RulesFunction`s, typically created using `bSync`, defining the sequence of steps for the thread.
 * @param repeat Optional configuration (`Repeat`) to control if and how the thread repeats its sequence of rules.
 * @returns A `RulesFunction` representing the combined behavior of the provided rules, potentially repeating.
 */
export type BThread = (rules: RulesFunction[], repeat?: Repeat) => RulesFunction

/**
 * Type guard to check if an unknown value conforms to the `BPEvent` structure.
 * Verifies if the value is an object with a `type` property that is a string.
 *
 * @param data The value to check.
 * @returns `true` if the value is a valid `BPEvent`, `false` otherwise.
 */
export const isBPEvent = (data: unknown): data is BPEvent => {
  return isTypeOf<{ [key: string]: unknown }>(data, 'object') && 'type' in data && isTypeOf<string>(data.type, 'string')
}

/**
 * Creates a behavioral thread (b-thread) by combining a sequence of synchronization rules.
 * A b-thread is a generator function (`RulesFunction`) that defines a strand of behavior
 * within a `bProgram`. It yields synchronization points (`Idioms`) to coordinate with other threads.
 *
 * @param rules An array of `RulesFunction`s, usually created with `bSync`, defining the steps of the thread.
 * @param repeat Controls if and how the thread repeats its sequence:
 *   - `true`: Repeats the entire sequence indefinitely.
 *   - `function`: Evaluated before each potential repetition; repeats if it returns `true`.
 *   - `undefined` or `false`: Executes the sequence only once.
 * @returns A `RulesFunction` representing the complete b-thread.
 * @example
 * // Thread that requests 'HOT' then 'COLD', and repeats forever.
 * const hotColdLoop = bThread(
 *   [
 *     bSync({ request: { type: 'HOT' } }),
 *     bSync({ request: { type: 'COLD' } })
 *   ],
 *   true // Repeat indefinitely
 * );
 *
 * // Thread that waits for 'START', then requests 'DONE' once.
 * const startThenDone = bThread(
 *   [
 *     bSync({ waitFor: 'START' }),
 *     bSync({ request: { type: 'DONE' } })
 *   ]
 *   // No repeat argument, runs once.
 * );
 */
export const bThread: BThread = (rules, repeat) => {
  return repeat ?
      function* () {
        while (isTypeOf<boolean>(repeat, 'boolean') ? repeat : repeat()) {
          const length = rules.length
          for (let i = 0; i < length; i++) {
            yield* rules[i]()
          }
        }
      }
    : function* () {
        const length = rules.length
        for (let i = 0; i < length; i++) {
          yield* rules[i]()
        }
      }
}

/**
 * Creates a single synchronization point (`RulesFunction`) for a b-thread.
 * This is the fundamental building block yielded by a b-thread's generator function.
 * It encapsulates the thread's intentions (`Idioms`) for a single step of the `bProgram`'s execution cycle.
 *
 * @template T The type of the event detail payload relevant to this synchronization point.
 * @param syncPoint The `Idioms<T>` object defining the thread's request, waitFor, block, and interrupt declarations for this step.
 * @returns A `RulesFunction` (a generator that yields the `syncPoint` once).
 * @example
 * // A synchronization point where the thread requests 'PING'.
 * const requestPing = bSync({ request: { type: 'PING' } });
 *
 * // A synchronization point where the thread waits for 'PONG' and blocks 'TIMEOUT'.
 * const waitPongBlockTimeout = bSync({ waitFor: 'PONG', block: 'TIMEOUT' });
 *
 * function* myThread(): RulesFunction {
 *   yield* requestPing();
 *   yield* waitPongBlockTimeout();
 * }
 */
export const bSync: BSync = (syncPoint) =>
  function* () {
    yield syncPoint
  }
