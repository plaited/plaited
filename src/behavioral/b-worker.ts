/**
 * @internal
 * @module define-worker
 *
 * Purpose: Enables behavioral programming within Web Worker contexts for background processing
 * Architecture: Adapts WorkerGlobalScope to support full bProgram lifecycle and event communication
 * Dependencies: b-program for core BP, get-public-trigger for security, get-plaited-trigger for lifecycle
 * Consumers: Worker script files that need behavioral programming capabilities
 *
 * Maintainer Notes:
 * - This is the worker-side counterpart to useWorker for main thread
 * - Creates a full behavioral program inside the worker thread
 * - Bidirectional event flow: main thread â†” worker via postMessage
 * - Public events filter incoming messages for security
 * - Automatic cleanup on worker termination via disconnectSet
 * - Async initialization supports dynamic handler setup
 *
 * Common modification scenarios:
 * - Supporting SharedWorker: Adapt message routing for multiple connections
 * - Adding initialization data: Pass config through first message
 * - Error boundaries: Wrap handlers in try-catch for resilience
 * - Performance monitoring: Track handler execution times
 *
 * Performance considerations:
 * - Worker initialization is async - may delay first message handling
 * - Message serialization overhead for all events
 * - DisconnectSet grows with each registered cleanup
 * - Public event filtering adds minimal overhead
 *
 * Known limitations:
 * - No access to DOM APIs in worker context
 * - Cannot share memory directly with main thread
 * - Transferable objects not supported in current implementation
 * - Worker must be in separate file for proper execution
 */
import {
  behavioral,
  type Behavioral,
  type Disconnect,
  type Handlers,
  type EventDetails,
  type BPEvent,
} from './behavioral.js'
import { getPublicTrigger } from './get-public-trigger.js'
import { getPlaitedTrigger } from './get-plaited-trigger.js'

/**
 * @internal
 * Extended arguments passed to the worker's bProgram function.
 * Includes worker-specific capabilities like send() for messaging main thread.
 */
type BProgramArgs = {
  send(data: BPEvent): void
  disconnect: Disconnect
} & Omit<ReturnType<Behavioral>, 'useFeedback'>
/**
 * Creates a behavioral program worker with type-safe message handling and lifecycle management.
 * Integrates Web Workers with Plaited's behavioral programming system for efficient background processing.
 *
 * @template A Type extending Handlers for event handling
 *
 * @param options Configuration object
 * @param options.bProgram Function defining worker behavior and event handlers
 * @param options.publicEvents Array of allowed event types for message filtering
 *
 * @example Image Processing Worker
 * ```tsx
 * // image-worker.ts
 * bWorker<{
 *   processImage: (args: { imageData: ImageData, filters: FilterOptions }) => void
 * }>({
 *   publicEvents: ['processImage'],
 *   bProgram({ send }) {
 *     return {
 *       processImage({ imageData, filters }) {
 *         // Process image in chunks to show progress
 *         const chunks = splitIntoChunks(imageData);
 *         chunks.forEach((chunk, index) => {
 *           const processed = applyFilters(chunk, filters);
 *
 *           // Report progress
 *           send({
 *             type: 'PROGRESS_UPDATE',
 *             detail: {
 *               progress: ((index + 1) / chunks.length) * 100,
 *               chunk: processed
 *             }
 *           });
 *         });
 *
 *         send({
 *           type: 'PROCESS_COMPLETE',
 *           detail: { success: true }
 *         });
 *       }
 *     };
 *   }
 * });
 *
 * // Usage in component:
 * const ImageEditor = bElement({
 *   tag: 'image-editor',
 *   shadowDom: (
 *     <div>
 *       <canvas p-target="canvas" />
 *       <div p-target="progress">0%</div>
 *       <button
 *         p-target="processBtn"
 *         p-trigger={{ click: 'START_PROCESSING' }}
 *       >
 *         Apply Filters
 *       </button>
 *     </div>
 *   ),
 *   bProgram({ $, trigger }) {
 *     const [canvas] = $<HTMLCanvasElement>('canvas');
 *     const [progress] = $('progress');
 *     const sendToWorker = useWorker(trigger, './image-worker.ts');
 *
 *     return {
 *       START_PROCESSING() {
 *         const ctx = canvas.getContext('2d');
 *         const imageData = ctx?.getImageData(0, 0, canvas.width, canvas.height);
 *
 *         sendToWorker({
 *           type: 'processImage',
 *           detail: {
 *             imageData,
 *             filters: { brightness: 1.2, contrast: 1.1 }
 *           }
 *         });
 *       },
 *
 *       PROGRESS_UPDATE({ detail }) {
 *         progress.render(`${Math.round(detail.progress)}%`);
 *         // Update canvas with processed chunk...
 *       },
 *
 *       PROCESS_COMPLETE() {
 *         progress.render('Processing complete!');
 *       }
 *     };
 *   }
 * });
 * ```
 *
 * @remarks
 * Worker Configuration:
 * - Must be defined in a separate file
 * - Runs in isolated thread context
 * - Has access to WorkerGlobalScope
 * - Cannot access DOM directly
 *
 * Message Handling:
 * - Type-safe message passing via generics
 * - Automatic event filtering based on publicEvents
 * - Built-in error handling
 * - Structured response format
 *
 * Best Practices:
 * 1. Worker Tasks:
 *    - CPU-intensive calculations
 *    - Data processing/analysis
 *    - Image/video processing
 *    - Complex algorithms
 *
 * 2. Communication:
 *    - Use typed messages
 *    - Send progress updates
 *    - Handle errors gracefully
 *    - Clean up resources
 *
 * 3. Performance:
 *    - Break large tasks into chunks
 *    - Report progress regularly
 *    - Consider data transfer costs
 *    - Use appropriate data structures
 */
export const bWorker = async <A extends EventDetails>({
  bProgram,
  publicEvents,
}: {
  bProgram: (args: BProgramArgs) => Handlers<A> | Promise<Handlers<A>>
  publicEvents: string[]
}) => {
  /**
   * @internal
   * Reference to WorkerGlobalScope for event handling.
   * Using 'self' directly can cause issues in some bundlers.
   */
  const context = self

  /**
   * @internal
   * Initialize a Behavioral Program instance within the worker.
   * This provides the core BP infrastructure for the worker thread.
   */
  const { useFeedback, trigger, ...rest } = behavioral()

  /**
   * @internal
   * Create filtered trigger that only accepts whitelisted events from main thread.
   * This prevents arbitrary event injection from compromising worker behavior.
   */
  const publicTrigger = getPublicTrigger({
    trigger,
    publicEvents,
  })

  /**
   * @internal
   * Message handler that forwards valid events from main thread to the behavioral program.
   * Type safety is enforced by publicTrigger's filtering.
   */
  const eventHandler = ({ data }: { data: BPEvent }) => publicTrigger(data)

  /**
   * @internal
   * Manages cleanup callbacks for proper resource disposal.
   * Pre-populated with message listener cleanup to prevent leaks.
   */
  const disconnectSet = new Set<Disconnect>()
  disconnectSet.add(() => {
    context.removeEventListener('message', eventHandler)
    disconnectSet.clear()
  })

  /**
   * @internal
   * Send function for communicating back to main thread.
   * Wraps postMessage for consistent event-based interface.
   */
  const send = (data: BPEvent) => context.postMessage(data)

  /**
   * @internal
   * Master disconnect that runs all cleanup callbacks and terminates worker.
   * Called on worker shutdown or explicit disconnect request.
   */
  const disconnect = () => {
    disconnectSet.forEach((disconnect) => void disconnect())
    self.close()
  }

  /**
   * @internal
   * Execute user-provided bProgram to get event handlers.
   * Await supports async initialization (e.g., loading resources).
   * Enhanced trigger enables automatic cleanup registration.
   */
  useFeedback(
    await bProgram({
      ...rest,
      send,
      disconnect,
      trigger: getPlaitedTrigger(trigger, disconnectSet),
    }),
  )

  /**
   * @internal
   * Start listening for messages from main thread.
   * This completes the bidirectional communication setup.
   */
  context.addEventListener('message', eventHandler, false)
}
