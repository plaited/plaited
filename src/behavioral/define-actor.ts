import { type BSync, type BThread, bThread, bSync } from './b-thread.js'
import {
  type EventDetails,
  type UseSnapshot,
  type BThreads,
  bProgram,
  type Disconnect,
  type Handlers,
} from './b-program.js'
import { getPlaitedTrigger, type PlaitedTrigger } from './get-plaited-trigger.js'
import { getPublicTrigger } from './get-public-trigger.js'

export const defineActor = <
  A extends EventDetails,
  C extends { [key: string]: unknown } = { [key: string]: unknown },
>(args: {
  publicEvents?: string[]
  bProgram: (
    args: {
      bSync: BSync
      bThread: BThread
      bThreads: BThreads
      disconnect: Disconnect
      trigger: PlaitedTrigger
      useSnapshot: UseSnapshot
    } & C,
  ) => Handlers<A> | Promise<Handlers<A>>
}) => {
  const { trigger, useFeedback, ...rest } = bProgram()
  const disconnectSet = new Set<Disconnect>()
  const disconnect = () => {
    disconnectSet.forEach((disconnect) => disconnect())
  }
  return async (ctx: C) => {
    const handlers = await args.bProgram({
      ...ctx,
      bSync,
      bThread,
      disconnect,
      trigger: getPlaitedTrigger(trigger, disconnectSet),
      ...rest,
    })
    useFeedback(handlers)
    return getPublicTrigger({ trigger, publicEvents: args?.publicEvents })
  }
}
