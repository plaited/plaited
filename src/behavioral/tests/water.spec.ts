import { test, expect } from 'bun:test'
import { behavioral, bSync, bThread, type SnapshotMessage } from 'plaited/behavioral'

/**
 * Test scenario: Demonstrates a basic behavioral program (`bProgram`).
 * It features a single b-thread (`addHot`) that sequentially requests the 'hot' event three times.
 * This showcases the fundamental concept of a thread making requests.
 *
 * Setup:
 * - A `bProgram` instance is created.
 * - A b-thread named 'addHot' is defined using `bThread` and `bSync`.
 *   - It consists of three steps, each requesting the 'hot' event.
 * - A feedback handler using `useFeedback` is registered to track when 'hot' events are selected.
 * - The program is initiated by triggering a 'start' event (though any event could start it).
 *
 * Expected Outcome:
 * - The 'hot' event handler should be called three times, in sequence.
 * - The `actual` array should contain ['hot', 'hot', 'hot'].
 */
test('Add hot water 3 times', () => {
  const actual: string[] = []
  const { bThreads, trigger, useFeedback } = behavioral()
  bThreads.set({
    addHot: bThread([
      bSync({ request: { type: 'hot' } }),
      bSync({ request: { type: 'hot' } }),
      bSync({ request: { type: 'hot' } }),
    ]),
  })
  useFeedback({
    hot() {
      actual.push('hot')
    },
  })
  trigger({ type: 'start' })
  expect(actual).toEqual(['hot', 'hot', 'hot'])
})

/**
 * Test scenario: Illustrates the behavior of multiple independent b-threads running concurrently.
 * One thread (`addHot`) requests 'hot' three times, while another (`addCold`) requests 'cold' three times.
 * This demonstrates the default event selection strategy based on thread registration order (priority).
 *
 * Setup:
 * - Similar to the previous test, but with an additional 'addCold' thread.
 * - Feedback handlers are registered for both 'hot' and 'cold' events.
 *
 * Expected Outcome:
 * - Since 'addHot' is registered first (implicitly higher priority), all its 'hot' requests
 *   are selected and executed before any 'cold' requests from 'addCold'.
 * - The `actual` array should contain ['hot', 'hot', 'hot', 'cold', 'cold', 'cold'].
 */
test('Add hot/cold water 3 times', () => {
  const actual: string[] = []
  const { bThreads, trigger, useFeedback } = behavioral()
  bThreads.set({
    addHot: bThread([
      bSync({ request: { type: 'hot' } }),
      bSync({ request: { type: 'hot' } }),
      bSync({ request: { type: 'hot' } }),
    ]),
    addCold: bThread([
      bSync({ request: { type: 'cold' } }),
      bSync({ request: { type: 'cold' } }),
      bSync({ request: { type: 'cold' } }),
    ]),
  })
  useFeedback({
    hot() {
      actual.push('hot')
    },
    cold() {
      actual.push('cold')
    },
  })
  trigger({ type: 'start' })
  expect(actual).toEqual(['hot', 'hot', 'hot', 'cold', 'cold', 'cold'])
})

/**
 * Test scenario: Demonstrates thread coordination using `waitFor` and `block`.
 * A third thread 'mixHotCold' is introduced to enforce alternation between 'hot' and 'cold' events.
 * It waits for 'hot' while blocking 'cold', then waits for 'cold' while blocking 'hot', repeating indefinitely.
 * This ensures the 'hot' and 'cold' events are interleaved.
 */
test('interleave', () => {
  const actual: string[] = []
  const { bThreads, trigger, useFeedback } = behavioral()
  bThreads.set({
    addHot: bThread([
      bSync({ request: { type: 'hot' } }),
      bSync({ request: { type: 'hot' } }),
      bSync({ request: { type: 'hot' } }),
    ]),
    addCold: bThread([
      bSync({ request: { type: 'cold' } }),
      bSync({ request: { type: 'cold' } }),
      bSync({ request: { type: 'cold' } }),
    ]),
    mixHotCold: bThread([bSync({ waitFor: 'hot', block: 'cold' }), bSync({ waitFor: 'cold', block: 'hot' })], true),
  })
  useFeedback({
    hot() {
      actual.push('hot')
    },
    cold() {
      actual.push('cold')
    },
  })
  trigger({ type: 'start' })
  expect(actual).toEqual(['hot', 'cold', 'hot', 'cold', 'hot', 'cold'])
})

/**
 * Test scenario: Demonstrates the use of `useSnapshot` to capture the state
 * of the behavioral program at each step (super-step).
 * This is useful for debugging and understanding the event selection process.
 * The captured snapshots are compared against a baseline snapshot.
 */
test('logging', () => {
  const snapshots: SnapshotMessage[] = []
  const { bThreads, trigger, useSnapshot } = behavioral()
  useSnapshot((snapshot: SnapshotMessage) => {
    snapshots.push(snapshot)
  })
  bThreads.set({
    addHot: bThread([
      bSync({ request: { type: 'hot' } }),
      bSync({ request: { type: 'hot' } }),
      bSync({ request: { type: 'hot' } }),
    ]),
    addCold: bThread([
      bSync({ request: { type: 'cold' } }),
      bSync({ request: { type: 'cold' } }),
      bSync({ request: { type: 'cold' } }),
    ]),
    mixHotCold: bThread([bSync({ waitFor: 'hot', block: 'cold' }), bSync({ waitFor: 'cold', block: 'hot' })], true),
  })
  trigger({ type: 'start' })
  expect(snapshots).toMatchSnapshot()
})
