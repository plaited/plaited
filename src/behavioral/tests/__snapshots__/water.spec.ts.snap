// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`logging 1`] = `
[
  [
    {
      "blockedBy": undefined,
      "detail": undefined,
      "interrupts": undefined,
      "priority": 0,
      "selected": true,
      "thread": "Symbol(start)",
      "trigger": true,
      "type": "start",
    },
    {
      "blockedBy": undefined,
      "detail": undefined,
      "interrupts": undefined,
      "priority": 1,
      "selected": false,
      "thread": "addHot",
      "trigger": false,
      "type": "hot",
    },
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "interrupts": undefined,
      "priority": 2,
      "selected": false,
      "thread": "addCold",
      "trigger": false,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": undefined,
      "detail": undefined,
      "interrupts": undefined,
      "priority": 1,
      "selected": true,
      "thread": "addHot",
      "trigger": false,
      "type": "hot",
    },
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "interrupts": undefined,
      "priority": 2,
      "selected": false,
      "thread": "addCold",
      "trigger": false,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "interrupts": undefined,
      "priority": 1,
      "selected": false,
      "thread": "addHot",
      "trigger": false,
      "type": "hot",
    },
    {
      "blockedBy": undefined,
      "detail": undefined,
      "interrupts": undefined,
      "priority": 2,
      "selected": true,
      "thread": "addCold",
      "trigger": false,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": undefined,
      "detail": undefined,
      "interrupts": undefined,
      "priority": 1,
      "selected": true,
      "thread": "addHot",
      "trigger": false,
      "type": "hot",
    },
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "interrupts": undefined,
      "priority": 2,
      "selected": false,
      "thread": "addCold",
      "trigger": false,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "interrupts": undefined,
      "priority": 1,
      "selected": false,
      "thread": "addHot",
      "trigger": false,
      "type": "hot",
    },
    {
      "blockedBy": undefined,
      "detail": undefined,
      "interrupts": undefined,
      "priority": 2,
      "selected": true,
      "thread": "addCold",
      "trigger": false,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": undefined,
      "detail": undefined,
      "interrupts": undefined,
      "priority": 1,
      "selected": true,
      "thread": "addHot",
      "trigger": false,
      "type": "hot",
    },
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "interrupts": undefined,
      "priority": 2,
      "selected": false,
      "thread": "addCold",
      "trigger": false,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": undefined,
      "detail": undefined,
      "interrupts": undefined,
      "priority": 2,
      "selected": true,
      "thread": "addCold",
      "trigger": false,
      "type": "cold",
    },
  ],
]
`;
